File: android-app/lib/core/assets/asset_allocation.dart
enum AssetAcquisitionCostSource {
  transactionAllocations,
  manual,
  manualUnknown,
}

extension AssetAcquisitionCostSourceX on AssetAcquisitionCostSource {
  String get storageKey => switch (this) {
        AssetAcquisitionCostSource.transactionAllocations =>
          'transaction_allocations',
        AssetAcquisitionCostSource.manual => 'manual',
        AssetAcquisitionCostSource.manualUnknown => 'manual_unknown',
      };

  static AssetAcquisitionCostSource fromStorage(String? value) =>
      switch (value) {
        'transaction_allocations' =>
          AssetAcquisitionCostSource.transactionAllocations,
        'manual_unknown' => AssetAcquisitionCostSource.manualUnknown,
        _ => AssetAcquisitionCostSource.manual,
      };
}

enum AssetAllocationCostQuality {
  exact,
  partial,
  pendingRefundAllocation,
  manualUnverified,
}

extension AssetAllocationCostQualityX on AssetAllocationCostQuality {
  String get storageKey => switch (this) {
        AssetAllocationCostQuality.exact => 'exact',
        AssetAllocationCostQuality.partial => 'partial',
        AssetAllocationCostQuality.pendingRefundAllocation =>
          'pending_refund_allocation',
        AssetAllocationCostQuality.manualUnverified => 'manual_unverified',
      };

  static AssetAllocationCostQuality fromStorage(String? value) =>
      switch (value) {
        'exact' => AssetAllocationCostQuality.exact,
        'pending_refund_allocation' =>
          AssetAllocationCostQuality.pendingRefundAllocation,
        'manual_unverified' => AssetAllocationCostQuality.manualUnverified,
        _ => AssetAllocationCostQuality.partial,
      };
}

class AssetAllocationLine {
  final int assetId;
  final int grossCents;
  final int refundCents;

  const AssetAllocationLine({
    required this.assetId,
    required this.grossCents,
    required this.refundCents,
  });

  int get netCents => grossCents - refundCents;
}

class AssetAllocationTotals {
  final int grossCents;
  final int refundCents;
  final int netCents;

  const AssetAllocationTotals({
    required this.grossCents,
    required this.refundCents,
    required this.netCents,
  });
}

class AssetAllocationPolicy {
  const AssetAllocationPolicy._();

  static AssetAllocationTotals validate({
    required int orderGrossCents,
    required int validOrderRefundCents,
    required Iterable<AssetAllocationLine> lines,
  }) {
    if (orderGrossCents < 0 ||
        validOrderRefundCents < 0 ||
        validOrderRefundCents > orderGrossCents) {
      throw ArgumentError('订单毛额或有效退款额不合法');
    }
    var gross = 0;
    var refund = 0;
    final assetIds = <int>{};
    for (final line in lines) {
      if (!assetIds.add(line.assetId)) {
        throw StateError('同一物品不能重复分配同一订单');
      }
      if (line.grossCents < 0 ||
          line.refundCents < 0 ||
          line.refundCents > line.grossCents) {
        throw ArgumentError('每件物品必须满足 0 <= 退款额 <= 毛额');
      }
      gross += line.grossCents;
      refund += line.refundCents;
    }
    if (gross > orderGrossCents) {
      throw StateError('物品毛额分配合计超过订单毛额');
    }
    if (refund > validOrderRefundCents) {
      throw StateError('物品退款分配合计超过订单有效退款');
    }
    final net = gross - refund;
    if (net > orderGrossCents - validOrderRefundCents) {
      throw StateError('物品净额分配合计超过订单净额');
    }
    return AssetAllocationTotals(
      grossCents: gross,
      refundCents: refund,
      netCents: net,
    );
  }
}

File: ios-app/QingJiCore/Sources/QingJiCore/Assets/AssetAllocation.swift
import Foundation

public enum AssetAcquisitionCostSource: String, Codable, CaseIterable, Hashable, Sendable {
    case transactionAllocations = "transaction_allocations"
    case manual
    case manualUnknown = "manual_unknown"
}

public enum AssetAllocationCostQuality: String, Codable, CaseIterable, Hashable, Sendable {
    case exact
    case partial
    case pendingRefundAllocation = "pending_refund_allocation"
    case manualUnverified = "manual_unverified"
}

public struct AssetAllocationLine: Equatable, Sendable {
    public let assetID: UUID
    public let grossCents: Int
    public let refundCents: Int

    public init(assetID: UUID, grossCents: Int, refundCents: Int) {
        self.assetID = assetID
        self.grossCents = grossCents
        self.refundCents = refundCents
    }

    public var netCents: Int { grossCents - refundCents }
}

public struct AssetAllocationTotals: Equatable, Sendable {
    public let grossCents: Int
    public let refundCents: Int
    public let netCents: Int

    public init(grossCents: Int, refundCents: Int, netCents: Int) {
        self.grossCents = grossCents
        self.refundCents = refundCents
        self.netCents = netCents
    }
}

public enum AssetAllocationPolicy {
    public static func validate(
        orderGrossCents: Int,
        validOrderRefundCents: Int,
        lines: some Sequence<AssetAllocationLine>
    ) throws -> AssetAllocationTotals {
        guard orderGrossCents >= 0,
              validOrderRefundCents >= 0,
              validOrderRefundCents <= orderGrossCents else {
            throw AssetAllocationError.invalidOrder
        }
        var gross = 0
        var refund = 0
        var assetIDs = Set<UUID>()
        for line in lines {
            guard assetIDs.insert(line.assetID).inserted else {
                throw AssetAllocationError.duplicateAsset
            }
            guard line.grossCents >= 0,
                  line.refundCents >= 0,
                  line.refundCents <= line.grossCents else {
                throw AssetAllocationError.invalidLine
            }
            gross += line.grossCents
            refund += line.refundCents
        }
        guard gross <= orderGrossCents else { throw AssetAllocationError.grossExceedsOrder }
        guard refund <= validOrderRefundCents else { throw AssetAllocationError.refundExceedsOrder }
        let net = gross - refund
        guard net <= orderGrossCents - validOrderRefundCents else {
            throw AssetAllocationError.netExceedsOrder
        }
        return AssetAllocationTotals(grossCents: gross, refundCents: refund, netCents: net)
    }
}

public enum AssetAllocationError: Error, Equatable, Sendable {
    case invalidOrder
    case duplicateAsset
    case invalidLine
    case grossExceedsOrder
    case refundExceedsOrder
    case netExceedsOrder
}
